### DISCUSSION TITLE
How to regist PageInfo in Schema

### DISCUSSION THREAD

**Discussion by ferrisyang** [Netflix/dgs-framework#371] [Q&A]
Use Relay style for pagination, and write PageInfo type in Schema,  will get error [The field type 'PageInfo' is not present] when start the program.
How to regist PageInfo type in Schema?

**Comment by ferrisyang** [comment]
I found a solution:

```java
@DgsScalar(name = "PageInfo")
public class PageInfoScalar implements Coercing<PageInfo, PageInfo> {

    private static final ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public PageInfo serialize(Object dataFetcherResult) throws CoercingSerializeException {
        if (dataFetcherResult instanceof PageInfo) {
            return (PageInfo) dataFetcherResult;
        }
        throw new CoercingSerializeException("Not a valid PageInfo");
    }

    @Override
    public PageInfo parseValue(Object input) throws CoercingParseValueException {
       return null;
    }

    @Override
    public PageInfo parseLiteral(Object input) throws CoercingParseLiteralException {
       return null;
    }
}
```

And in schema, add `PageInfo` scalar definition:

```schema
scalar PageInfo
```

**Reply by sbilello** [reply]
https://github.com/Netflix/dgs-codegen/issues/141#issuecomment-853178200 @srinivasankavitha so we need to register as scalar?

**Reply by srinivasankavitha** [reply]
If you want to use it as a scalar type, then sure, but I haven't used or had to use that pattern. It should work without having to declare it as a scalar as well as I had pointed out in the example.

**Reply by sbilello** [reply]
It does not work without a `Scalar` registration. Please check this branch to replicate the bug.
https://github.com/sbilello/dgs-examples-java/commit/16f46d595f23e0c3488dcd3d6c1e85829ecd8fa5
Thanks
@srinivasankavitha 

```
***************************
APPLICATION FAILED TO START
***************************

Description:

There are problems with the GraphQL Schema:
	 * The field type 'PageInfo' is not present when resolving type 'Show' [@15:1]
```

**Reply by sbilello** [reply]
I created this PR https://github.com/Netflix/dgs-examples-java/pull/30/files that shows the problem and how it can work with the current framework.

**Reply by amondnet** [reply]
@sbilello @ferrisyang 
`PageInfo` is not scalar, you should mapping `PageInfo` to `graphql.relay.PageInfo` ([source](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/relay/PageInfo.java)).
https://netflix.github.io/dgs/generating-code-from-schema/#configuring-code-generation

```kotlin
//build.gradle.kts
@OptIn(kotlin.ExperimentalStdlibApi::class)
tasks.withType<com.netflix.graphql.dgs.codegen.gradle.GenerateJavaTask> {
  generateClient = true
  typeMapping = mutableMapOf<String, String>(
    //"JSON" to "graphql.scalars.object.JsonScalar",
    //"Object" to "graphql.scalars.object.ObjectScalar",
    //"Long" to "kotlin.Long",
    "PageInfo" to "graphql.relay.PageInfo",
    "Edge" to "graphql.relay.Edge",
    "Connection" to "graphql.relay.Connection"
  )
}
```

**Reply by sbilello** [reply]
@amondnet I solved the problem over here: It looks like that the problem is on defining the schema: https://github.com/Netflix/dgs-examples-java/pull/30#issuecomment-860846334
Do you think that it can work with just the mapping without specifying the `type PageInfo` in the schema?

**Comment by amondnet** [comment]
You should mapping `PageInfo` to `graphql.relay.PageInfo` ([source](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/relay/PageInfo.java)).
https://netflix.github.io/dgs/generating-code-from-schema/#configuring-code-generation

```kotlin
//build.gradle.kts
@OptIn(kotlin.ExperimentalStdlibApi::class)
tasks.withType<com.netflix.graphql.dgs.codegen.gradle.GenerateJavaTask> {
  generateClient = true
  typeMapping = mutableMapOf<String, String>(
    //"JSON" to "graphql.scalars.object.JsonScalar",
    //"Object" to "graphql.scalars.object.ObjectScalar",
    //"Long" to "kotlin.Long",
    "PageInfo" to "graphql.relay.PageInfo",
    "Edge" to "graphql.relay.Edge",
    "Connection" to "graphql.relay.Connection"
  )
}
```
